backends/p4tools/README.md

```diff
@@ -0,0 +1,110 @@
+# P4Tools - Testing Tools For P4 Targets
+
+## Directory Structure
+
+```
+p4tools
+ ├─ cmake       ── common CMake modules
+ ├─ common      ── C++ source: common code for the various components of p4check
+ ├─ mutate      ── C++ source: p4mutate
+ ├─ p4check     ── C++ source: the main entry point for p4check
+ ├─ scripts
+ │   ├─ hooks   ── git hooks
+ │   └─ tools   ── development-support scripts (e.g., cpplint)
+ ├─ smith       ── C++ source: p4smith
+ ├─ submodules  ── external dependencies
+ └─ testgen     ── C++ source: p4testgen
+```
+
+## P4Tools
+P4Tools is a collection of tools that make testing P4 targets and programs a little easier. So far the platform supports the following tools and projects:
+
+- [P4Testgen](https://github.com/p4lang/p4c/tree/main/backends/p4tools/testgen): An input-output test case generator for P4.
+
+## Building
+
+P4Tools can be built using the following CMAKE configuration in the P4C repository.
+
+```
+mkdir build
+cd build
+cmake .. -DENABLE_GMP=OFF -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DENABLE_TOOLS=ON
+make
+```
+
+## Dependencies
+* [inja](https://github.com/pantor/inja) template engine for testcase generation.
+* [gsl-lite](https://github.com/gsl-lite/gsl-lite) C++ Core Guidelines Support Library, used to
+support and enforce C++ best practices.
+
+## Development Style
+Currently, each C++ source directory has a few subdirectories, including:
+* `core`, containing the core functionality of the submodule; and
+* `lib`, containing supporting data structures.
+
+The distinction between the two can be fuzzy. Here are some guiding principles
+for where to find/put what:
+* If it depends on anything in `core`, it belongs in `core`.
+* If it's something that resembles a general-purpose data structure (e.g., an
+  environment or a symbol table), it's probably in `lib`.
+
+### C++ Coding style
+
+We generally follow the [Google C++ Style
+Guide](https://google.github.io/styleguide/cppguide.html). This is partially
+enforced by `cpplint` and `clang-format` and their respective configuration files. Both tools run in a git hook and as part of CI. To run `clang-format` on Ubuntu 16.04, install it with `pip3 install --user clang-format`.
+
+Some deviations from the Style Guide are highlighted below. While this
+repository depends heavily on the main P4 compiler ([P4C](https://github.com/p4lang/p4c/)), the code there doesn't always follow our style; don't follow its precedent!
```

Comment:
that is unfortunate and too late to change now.

Reply:
Actually, I do not think the differences are too big. However, we have more automated tools to check for style, autoformatting, etc. For example, we run https://github.com/fruffy-bfn/p4c/blob/p4tools/backends/p4tools/.clang-format for the PRs to automatically check code formatting. We can do this for the entire compiler. It would just require one big commit that autoformats the code.

Similarly, we run [IWYU](https://github.com/include-what-you-use/include-what-you-use) to clean up all the includes.